hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiServiceImpl.java

```diff
@@ -0,0 +1,387 @@
+
+    // No need to "volatile" since this field is guarded and happens-before is handled by
+    // `synchronized` blocks of `serviceMutex` instance.
+    private ServiceRegistration serviceRegistration;
+
+    // Defined as `volatile` because even though this field is written inside `synchronized` blocks,
+    // it can be read from any thread without `synchronized` blocks and without `volatile` they may seen invalid value.
+    private volatile HazelcastOSGiInstance hazelcastInstance;
+
+    // Defined as `volatile` because even though this field is updated inside `synchronized` blocks,
+    // it can be read from any thread without `synchronized` blocks and without `volatile` they may seen invalid value.
+    private volatile boolean active;
+
+    public HazelcastOSGiServiceImpl(Bundle ownerBundle) {
+        this(ownerBundle, DEFAULT_ID);
+    }
+
+    public HazelcastOSGiServiceImpl(Bundle ownerBundle, String id) {
+        this.ownerBundle = ownerBundle;
+        this.ownerBundleContext = ownerBundle.getBundleContext();
+        this.id = id;
+    }
+
+    private void checkActive() {
+        if (!active) {
+            throw new IllegalStateException("Hazelcast OSGI Service is not active!");
+        }
+    }
+
+    private Config getConfig(Config config) {
+        if (config == null) {
+            config = new XmlConfigBuilder().build();
+        }
+        GroupConfig groupConfig = config.getGroupConfig();
+        if (groupConfig == null
+                || StringUtil.isNullOrEmpty(groupConfig.getName())
+                || GroupConfig.DEFAULT_GROUP_NAME.equals(groupConfig.getName())) {
+            if (!Boolean.getBoolean(HAZELCAST_OSGI_GROUPING_DISABLED)) {
+                String groupName = id;
+                if (groupConfig == null) {
+                    config.setGroupConfig(new GroupConfig(groupName));
+                } else {
+                    groupConfig.setName(groupName);
+                }
+            }
+        }
+        return config;
+    }
+
+    private HazelcastInstance createHazelcastInstance(Config config) {
+        return Hazelcast.newHazelcastInstance(getConfig(config));
+    }
+
+    private HazelcastOSGiInstance registerInstance(HazelcastInstance instance) {
+        HazelcastOSGiInstance hazelcastOSGiInstance;
+        if (instance instanceof HazelcastOSGiInstance) {
+            hazelcastOSGiInstance = (HazelcastOSGiInstance) instance;
+        } else {
+            hazelcastOSGiInstance = new HazelcastOSGiInstanceImpl(instance, this);
+        }
+        ServiceRegistration serviceRegistration;
+        if (!Boolean.getBoolean(HAZELCAST_OSGI_REGISTER_DISABLED)) {
+            serviceRegistration =
+                    ownerBundleContext.registerService(HazelcastInstance.class.getName(), hazelcastOSGiInstance, null);
+        } else {
+            serviceRegistration = EMPTY_SERVICE_REGISTRATION;
+        }
+        instanceServiceRegistrationMap.put(hazelcastOSGiInstance, serviceRegistration);
```

Comment:
Do we need to add `try catch`  block here, maybe if service registration fails, we should not put entries to `instanceServiceRegistrationMap`,`instanceMap `.